// @nrwl/node/src/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBaseWebpackPartial = exports.OUT_FILENAME = void 0;
const ts = require("typescript");
const license_webpack_plugin_1 = require("license-webpack-plugin");
const CircularDependencyPlugin = require("circular-dependency-plugin");
const tsconfig_paths_webpack_plugin_1 = require("tsconfig-paths-webpack-plugin");
const typescript_1 = require("@nrwl/workspace/src/utilities/typescript");
const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
exports.OUT_FILENAME = 'main.js';
function getBaseWebpackPartial(options) {
    // TODO(jack): Remove in Nx 13
    const { CopyWebpackPlugin, webpack } = require('../webpack/entry');
    const { options: compilerOptions } = typescript_1.readTsConfig(options.tsConfig);
    const supportsEs2015 = compilerOptions.target !== ts.ScriptTarget.ES3 &&
        compilerOptions.target !== ts.ScriptTarget.ES5;
    const mainFields = [...(supportsEs2015 ? ['es2015'] : []), 'module', 'main'];
    const extensions = ['.ts', '.tsx', '.mjs', '.js', '.jsx'];
    const webpackConfig = {
        entry: {
            main: [options.main],
        },
        devtool: options.sourceMap ? 'source-map' : false,
        mode: options.optimization ? 'production' : 'development',
        output: {
            path: options.outputPath,
            filename: exports.OUT_FILENAME,
        },
        module: {
            rules: [
                {
                    test: /\.([jt])sx?$/,
                    loader: require.resolve(`ts-loader`),
                    exclude: /node_modules/,
                    options: {
                        configFile: options.tsConfig,
                        transpileOnly: true,
                        // https://github.com/TypeStrong/ts-loader/pull/685
                        experimentalWatchApi: true,
                    },
                },
            ],
        },
        resolve: {
            extensions,
            alias: getAliases(options),
            plugins: [
                new tsconfig_paths_webpack_plugin_1.default({
                    configFile: options.tsConfig,
                    extensions,
                    mainFields,
                }),
            ],
            mainFields,
        },
        performance: {
            hints: false,
        },
        plugins: [
            new ForkTsCheckerWebpackPlugin({
                typescript: {
                    enabled: true,
                    configFile: options.tsConfig,
                    memoryLimit: options.memoryLimit || 2018,
                },
            }),
        ],
        watch: options.watch,
        watchOptions: {
            poll: options.poll,
        },
        stats: getStatsConfig(options),
    };
    const extraPlugins = [];
    if (options.progress) {
        extraPlugins.push(new webpack.ProgressPlugin());
    }
    if (options.extractLicenses) {
        extraPlugins.push(new license_webpack_plugin_1.LicenseWebpackPlugin({
            stats: {
                errors: false,
            },
            perChunkOutput: false,
            outputFilename: `3rdpartylicenses.txt`,
        }));
    }
    // process asset entries
    if (Array.isArray(options.assets) && options.assets.length > 0) {
        const copyWebpackPluginInstance = new CopyWebpackPlugin({
            patterns: options.assets.map((asset) => {
                var _a;
                return {
                    context: asset.input,
                    // Now we remove starting slash to make Webpack place it from the output root.
                    to: asset.output,
                    from: asset.glob,
                    globOptions: {
                        ignore: [
                            '.gitkeep',
                            '**/.DS_Store',
                            '**/Thumbs.db',
                            ...((_a = asset.ignore) !== null && _a !== void 0 ? _a : []),
                        ],
                        dot: true,
                    },
                };
            }),
        });
        new CopyWebpackPlugin({
            patterns: options.assets.map((asset) => {
                var _a;
                return {
                    context: asset.input,
                    // Now we remove starting slash to make Webpack place it from the output root.
                    to: asset.output,
                    from: asset.glob,
                    globOptions: {
                        ignore: [
                            '.gitkeep',
                            '**/.DS_Store',
                            '**/Thumbs.db',
                            ...((_a = asset.ignore) !== null && _a !== void 0 ? _a : []),
                        ],
                        dot: true,
                    },
                };
            }),
        });
        extraPlugins.push(copyWebpackPluginInstance);
    }
    if (options.showCircularDependencies) {
        extraPlugins.push(new CircularDependencyPlugin({
            exclude: /[\\\/]node_modules[\\\/]/,
        }));
    }
    webpackConfig.plugins = [...webpackConfig.plugins, ...extraPlugins];
    return webpackConfig;
}
exports.getBaseWebpackPartial = getBaseWebpackPartial;
function getAliases(options) {
    return options.fileReplacements.reduce((aliases, replacement) => (Object.assign(Object.assign({}, aliases), { [replacement.replace]: replacement.with })), {});
}
// TODO(jack): Update the typing with new version of webpack -- was returning Stats.ToStringOptions in webpack 4
// The StatsOptions type needs to be exported from webpack
// PR: https://github.com/webpack/webpack/pull/12875
function getStatsConfig(options) {
    return {
        hash: true,
        timings: false,
        cached: false,
        cachedAssets: false,
        modules: false,
        warnings: true,
        errors: true,
        colors: !options.verbose && !options.statsJson,
        chunks: !options.verbose,
        assets: !!options.verbose,
        chunkOrigins: !!options.verbose,
        chunkModules: !!options.verbose,
        children: !!options.verbose,
        reasons: !!options.verbose,
        version: !!options.verbose,
        errorDetails: !!options.verbose,
        moduleTrace: !!options.verbose,
        usedExports: !!options.verbose,
    };
}
