// @nrwl/node/src/utils/node.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNodeWebpackConfig = void 0;
const app_root_1 = require("@nrwl/tao/src/utils/app-root");
const config_1 = require("./config");
function getNodePartial(options) {
    const { nodeExternals } = require('../webpack/entry');
    const webpackConfig = {
        output: {
            libraryTarget: 'commonjs',
        },
        target: 'node',
        node: false,
    };
    if (options.optimization) {
        webpackConfig.optimization = {
            minimize: false,
            concatenateModules: false,
        };
    }
    if (options.externalDependencies === 'all') {
        const modulesDir = `${app_root_1.appRootPath}/node_modules`;
        webpackConfig.externals = [nodeExternals({ modulesDir })];
    }
    else if (Array.isArray(options.externalDependencies)) {
        webpackConfig.externals = [
            function (context, request, callback) {
                if (options.externalDependencies.includes(request)) {
                    // not bundled
                    return callback(null, `commonjs ${request}`);
                }
                // bundled
                callback();
            },
        ];
    }
    return webpackConfig;
}
function getNodeWebpackConfig(options) {
    const { webpackMerge } = require('../webpack/entry');
    return webpackMerge([
        config_1.getBaseWebpackPartial(options),
        getNodePartial(options),
    ]);
}
exports.getNodeWebpackConfig = getNodeWebpackConfig;
